overlord/snapstate/refresh.go

```diff
@@ -0,0 +1,254 @@
+// downloading or stopping services for the update, we can check that only
+// services are running, that is, that no non-service apps or hooks are
+// currently running.
+//
+// Since services are stopped during the update this provides a good early
+// precondition check.  The check is also deliberately racy, both at the level
+// of processes forking and exiting and at the level of snap-confine launching
+// new commands. The hard check needs to be synchronized but the soft check
+// doesn't require this since it would serve no purpose. After the soft check
+// passes the user is free to start snap applications and block the hard check.
+func SoftRefreshCheck(info *snap.Info) error {
+	// Find the set of PIDs that belong to the snap, excluding any that belong
+	// to services since services are stopped for refresh and will likely not
+	// interfere.
+	snapName := info.SnapName()
+	pidSet, err := pidSetOfSnap(snapName)
+	if err != nil {
+		return err
+	}
+	for _, app := range info.Apps {
+		if app.IsService() {
+			pids, err := pidsOfSecurityTag(app.SecurityTag())
+			if err != nil {
+				return err
+			}
+			for _, pid := range pids {
+				delete(pidSet, pid)
+			}
+		}
+	}
+	if len(pidSet) == 0 {
+		return nil
+	}
+
+	// Some PIDs belong to non-service applications. Let's find out what we
+	// can and produce an informative error.
+	pids := make([]int, 0, len(pidSet))
+	for pid := range pidSet {
+		pids = append(pids, pid)
+	}
+	sort.Ints(pids)
+
+	var busyAppNames []string
+	for name, app := range info.Apps {
+		// Skip services since those are filtered out above.
+		if app.IsService() {
+			continue
+		}
+		isBusy, err := isSecurityTagBusy(app.SecurityTag())
+		if err != nil {
+			return err
+		}
+		if isBusy {
+			busyAppNames = append(busyAppNames, name)
+		}
+	}
+	sort.Strings(busyAppNames)
+
+	var busyHookNames []string
+	for name, hook := range info.Hooks {
+		isBusy, err := isSecurityTagBusy(hook.SecurityTag())
+		if err != nil {
+			return err
+		}
+		if isBusy {
+			busyHookNames = append(busyHookNames, name)
+		}
+	}
+	sort.Strings(busyHookNames)
+
+	return &BusySnapError{
+		pids:          pids,
+		snapName:      snapName,
+		busyAppNames:  busyAppNames,
+		busyHookNames: busyHookNames,
+	}
+}
+
+// HardRefreshCheck looks if there are any processes alive.
+//
+// The check is designed to run late in the refresh pipeline, after stopping
+// snap services. At this point services should be stopped, hooks should no
+// longer run, and applications should be barred from running externally (e.g.
+// by grabbing the per-snap lock around that phase of the update).
+//
+// The check looks at the set of PIDs in the freezer cgroup associated with a
+// given snap. Presence of any processes indicates that a snap is busy and
+// refresh cannot proceed.
+func HardRefreshCheck(snapName string) error {
+	pidSet, err := pidSetOfSnap(snapName)
+	if err != nil {
+		return err
+	}
+	if len(pidSet) > 0 {
```

Comment:
even for the hard check we should ignore services that have refresh-mode: endure,  the snap is asking us explicitly to let them run, not much we can do

Reply:
Ah, this is interesting. So the hard check should still ignore services with endure mode? I guess we don't have many of those but this still means that the original problem remains for a subset of our users.

Reply:
Endure mode services are now handled.